File: src/investments/locked-in-retirement-account.ts
/*
Revised
    2025-08-12
*/

import { ByJurisdiction, FederalCode, ProvinceCode } from '../misc';

export interface ConversionRule {
    minimumAge: number;
    maximumAge: number;
}

export const MAX_CONVERSION_AGE = 71;

const federalConversionRule: ConversionRule = {
    minimumAge: 55,
    maximumAge: MAX_CONVERSION_AGE,
};

const conversionRules: ByJurisdiction<ConversionRule | null> = {
    CA: federalConversionRule,
    AB: { minimumAge: 50, maximumAge: MAX_CONVERSION_AGE },
    BC: { minimumAge: 50, maximumAge: MAX_CONVERSION_AGE },
    MB: { minimumAge: 0, maximumAge: MAX_CONVERSION_AGE },
    NB: { minimumAge: 0, maximumAge: MAX_CONVERSION_AGE },
    NL: { minimumAge: 55, maximumAge: MAX_CONVERSION_AGE },
    NS: { minimumAge: 55, maximumAge: MAX_CONVERSION_AGE },
    NT: federalConversionRule,
    NU: federalConversionRule,
    ON: { minimumAge: 55, maximumAge: MAX_CONVERSION_AGE },
    PE: null,
    QC: { minimumAge: 55, maximumAge: MAX_CONVERSION_AGE },
    SK: { minimumAge: 55, maximumAge: MAX_CONVERSION_AGE },
    YT: federalConversionRule,
};

const federalUnlockingPct: number = 0.50;

const unlockingPct: ByJurisdiction<number | null> = {
    CA: federalUnlockingPct,
    AB: 0.50,
    BC: null,
    MB: 1.00,
    NB: 0.50,
    NL: 0.50,
    NS: 0.50,
    NT: federalUnlockingPct,
    NU: federalUnlockingPct,
    ON: 0.50,
    PE: null,
    QC: null,
    SK: 1.00,
    YT: federalUnlockingPct,
};

export const getUnlockingPct = (
    jurisdiction: ProvinceCode | FederalCode,
): number | null => unlockingPct[jurisdiction];

export const getConversionRules = (
    jurisdiction: ProvinceCode | FederalCode,
): ConversionRule | null => conversionRules[jurisdiction];

export const canUnlock = (jurisdiction: ProvinceCode | FederalCode): boolean => {
    const pct = getUnlockingPct(jurisdiction);
    if (!pct) return false;
    return pct > 0;
};

export const canConvert = (jurisdiction: ProvinceCode | FederalCode, age: number): boolean => {
    const rule = getConversionRules(jurisdiction);
    if (!rule) return false;
    return age >= rule.minimumAge && age <= rule.maximumAge;
};

File: src/commonMain/kotlin/com/equisoft/taxca/investments/LockedInRetirementAccount.kt
/*
Revised
    2025-08-12
*/

package com.equisoft.taxca.investments

import com.equisoft.taxca.misc.Jurisdiction

data class ConversionRule(
    val minimumAge: Int,
    val maximumAge: Int,
)

const val MaxConversionAge: Int = 71

private val federalConversionRule: ConversionRule = ConversionRule(
    minimumAge = 55,
    maximumAge = MaxConversionAge,
)

private val conversionRules: Map<Jurisdiction, ConversionRule?> = mapOf(
    Jurisdiction.CA to federalConversionRule,
    Jurisdiction.AB to ConversionRule(minimumAge = 50, maximumAge = MaxConversionAge),
    Jurisdiction.BC to ConversionRule(minimumAge = 50, maximumAge = MaxConversionAge),
    Jurisdiction.MB to ConversionRule(minimumAge = 0, maximumAge = MaxConversionAge),
    Jurisdiction.NB to ConversionRule(minimumAge = 0, maximumAge = MaxConversionAge),
    Jurisdiction.NL to ConversionRule(minimumAge = 55, maximumAge = MaxConversionAge),
    Jurisdiction.NS to ConversionRule(minimumAge = 55, maximumAge = MaxConversionAge),
    Jurisdiction.NT to federalConversionRule,
    Jurisdiction.NU to federalConversionRule,
    Jurisdiction.ON to ConversionRule(minimumAge = 55, maximumAge = MaxConversionAge),
    Jurisdiction.PE to null,
    Jurisdiction.QC to ConversionRule(minimumAge = 55, maximumAge = MaxConversionAge),
    Jurisdiction.SK to ConversionRule(minimumAge = 55, maximumAge = MaxConversionAge),
    Jurisdiction.YT to federalConversionRule,
)

private const val federalUnlockingPct: Double = 0.50

private val unlockingPct: Map<Jurisdiction, Double?> = mapOf(
    Jurisdiction.CA to federalUnlockingPct,
    Jurisdiction.AB to 0.50,
    Jurisdiction.BC to null,
    Jurisdiction.MB to 1.00,
    Jurisdiction.NB to 0.50,
    Jurisdiction.NL to 0.50,
    Jurisdiction.NS to 0.50,
    Jurisdiction.NT to federalUnlockingPct,
    Jurisdiction.NU to federalUnlockingPct,
    Jurisdiction.ON to 0.50,
    Jurisdiction.PE to null,
    Jurisdiction.QC to null,
    Jurisdiction.SK to 1.00,
    Jurisdiction.YT to federalUnlockingPct,
)

fun getUnlockingPct(jurisdiction: Jurisdiction): Double? = unlockingPct[jurisdiction]

fun getConversionRules(jurisdiction: Jurisdiction): ConversionRule? = conversionRules[jurisdiction]

fun canUnlock(jurisdiction: Jurisdiction): Boolean {
    // Legacy JS: `if (!pct) return false` — falsy for both null and 0.
    val pct = getUnlockingPct(jurisdiction)
    if (pct == null || pct == 0.0) return false
    return pct > 0
}

fun canConvert(jurisdiction: Jurisdiction, age: Int): Boolean {
    val rule = getConversionRules(jurisdiction) ?: return false
    return age >= rule.minimumAge && age <= rule.maximumAge
}
